adapters/amqp-vertx-base/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -615,7 +615,23 @@ protected void handleRemoteReceiverOpen(final ProtonConnection conn, final Proto
 
                     spanPreparationFuture
                             .compose(ar -> onMessageReceived(ctx)
-                            .onComplete(r -> msgSpan.finish()));
+                                    .onSuccess(ok -> msgSpan.finish())
+                                    .onFailure(error -> Optional.ofNullable(ctx.getAddress())
+                                            .ifPresentOrElse(address -> isTerminalError(
+                                                    error,
+                                                    address.getTenantId(),
+                                                    address.getResourceId(),
+                                                    ctx.getAuthenticatedDevice(),
+                                                    msgSpan.context())
+                                                            .map(isTerminalError -> {
+                                                                span.log("closing connection to device ["
+                                                                        + ctx.getAddress().getResourceId() + "]");
+                                                                conn.close();
+                                                                conn.disconnect();
+                                                                return null;
+                                                            })
```

Comment:
The `if(isTerminalError)` check is missing here.

I think the whole block, starting with `Optional.ofNullable(ctx.getAddress()).ifPresentOrElse` can be made much more readable by introducing an extra 
````
private Future<Boolean> isTerminalError(final Throwable error, final AmqpContext context, final SpanContext spanContext)
````
method, doing the `ctx.getAddress()` null check in there (either returning `Future.failedFuture(..)` or just `Future.succeededFuture(false)` then).

And the above `.map(isTerminalError -> {` can be replaced with `.onSuccess(isTerminalError -> {`, skipping the `return null;` at the end.

Reply:
I have extracted into a new method and incorporated other changes too. Would you mind taking a look again?